core/services/ocr2/plugins/ccip/cache/tokenpool.go

```
@@ -0,0 +1,94 @@
+) *CachedChain[map[common.Address]common.Address] {
+	return &CachedChain[map[common.Address]common.Address]{
+		observedEvents: []common.Hash{
+			abihelpers.EventSignatures.FeeTokenAdded,
+			abihelpers.EventSignatures.FeeTokenRemoved,
+			abihelpers.EventSignatures.PoolAdded,
+			abihelpers.EventSignatures.PoolRemoved,
+		},
+		logPoller:               lp,
+		address:                 []common.Address{priceRegistry.Address(), offRamp.Address()},
+		optimisticConfirmations: optimisticConfirmations,
+		lock:                    &sync.RWMutex{},
+		value:                   make(map[common.Address]common.Address),
+		lastChangeBlock:         0,
+		origin:                  newTokenPoolsOrigin(lggr, offRamp, priceRegistry),
+	}
+}
+
+func newTokenPoolsOrigin(lggr logger.Logger, offRamp evm_2_evm_offramp.EVM2EVMOffRampInterface, priceReg price_registry.PriceRegistryInterface) *tokenPools {
+	return &tokenPools{
+		lggr:          lggr,
+		offRamp:       offRamp,
+		priceRegistry: priceReg,
+	}
+}
+
+type tokenPools struct {
+	lggr          logger.Logger
+	offRamp       evm_2_evm_offramp.EVM2EVMOffRampInterface
+	priceRegistry price_registry.PriceRegistryInterface
+}
+
+func (t *tokenPools) Copy(value map[common.Address]common.Address) map[common.Address]common.Address {
+	return copyMap(value)
+}
+
+func (t *tokenPools) CallOrigin(ctx context.Context) (map[common.Address]common.Address, error) {
+	const numWorkers = 5
+
+	destTokens, err := getDestinationAndFeeTokens(ctx, t.offRamp, t.priceRegistry)
+	if err != nil {
+		return nil, err
+	}
+
+	eg := new(errgroup.Group)
+	eg.SetLimit(numWorkers)
+	var mu sync.Mutex
+
+	mapping := make(map[common.Address]common.Address, len(destTokens))
+	for _, token := range destTokens {
+		token := token
+		eg.Go(func() error {
+			poolAddress, err := t.offRamp.GetPoolByDestToken(&bind.CallOpts{Context: ctx}, token)
+			if err != nil {
+				t.lggr.Infow("get token pool for token '%s': %w", token, err)
+				return nil
```

interesting - is the intention a best effort update of the cache? Seems much better to actually return an error, then the next call to Get will retry

---

if the code looks like this
```
tokens = getDestinationAndFeeTokens()
for token in tokens:
   pool, err = offramp.getpool(token)
   if err != nil -> return err
```

and for whatever reason offRamp *always* returns an error for a specific token, e.g. token pool is not configured
then the whole observation round would die until the issue for the specific token is fixed. I want to avoid this, that's why I wrote it in such way.

Single point of failure should not be a token config for the observation round.

Ideally we want a cache that is able to fetch specific keys and do error handling on the plugin level.

```
// plugin
for token in tokens:
   decimals, err = cache.getDecimals(ctx, someAddress)
   if err != nil -> ...
```

---

> e.g. token pool is not configured

I don't think this is a scenario we want graceful recovery. A one-sided token configuration is a critical failure which should never happen and if it does stopping observations is appropriate

---

Ok makes sense then to return an error and stop the observation round.

---

After applying this change the smoke tests are failing.
@AnieeG are the token pools in the smoke tests properly configured? The `getDestinationAndFeeTokens` returns some token that returns error for `r.config.offRamp.GetPoolByDestToken` @connorwstein this is exactly the scenario I wanted to prevent from happening on prod.

---

Sounds like we discovered a test misconfiguration that was being masked by this code. Its prevented in prod by the config checker

---

Hey @dimkouv @connorwstein I am not providing any dest tokens into offramp [constructor](https://github.com/smartcontractkit/ccip/blob/ccip-develop/integration-tests/ccip-tests/actions/ccip_helpers.go#L1066-L1069). I am providing source tokens and corresponding dest pool. Where should I pass the dest tokens into offramp?

---

dmed